### ISSUE TITLE
BSP MEM POSIX realloc impl NULL input ptr handling, diff. vs. C standard

### ISSUE THREAD

**Issue by atigyi** [GoogleCloudPlatform/iot-device-sdk-embedded-c#125]
The SDK BSP MEM's `realloc` implementation does not enforce C standard behaviour. This is required if the underlying platform's `realloc` function doesn't meet the standard. In this particular case the behaviour on NULL input ptr value differs from the standard (below) for Raspberry Pi / ARM architecture.

[iot-device-sdk's POSIX realloc function impl.](https://github.com/GoogleCloudPlatform/iot-device-sdk-embedded-c/blob/master/src/bsp/platform/posix/iotc_bsp_mem_posix.c#L24-L26)

**realloc - memory reallocator** : *"If ptr is a null pointer, realloc() shall be equivalent to malloc() for the specified size."*
[C `realloc` C standard definition.](https://pubs.opengroup.org/onlinepubs/009695399/functions/realloc.html)

**Some more details**
- Seems "only" the Wolfssl uses `realloc` and/or uses 'realloc' with NULL ptr input.
- The SDK actually propagates the standard C adherence to the underlying platform's `realloc` function by only forwarding the parameters intact, without applying any logic.

**Comment by sridharan-8888** [comment]
Additional info on the issue, it seems to be related to 'memory_limitter'
- Device: Raspberry PI
- OS : Raspien/Raspberry PI OS
With the 'memory_limitter' debug extension
Code causing the issue:
https://github.com/GoogleCloudPlatform/iot-device-sdk-embedded-c/blob/master/src/libiotc/debug_extensions/memory_limiter/iotc_memory_limiter.c#L230-L232

**Comment by atigyi** [comment]
Indeed, this isn't C standard compatible. Did you conduct tests with [if-null-return-null lines](https://github.com/GoogleCloudPlatform/iot-device-sdk-embedded-c/blob/master/src/libiotc/debug_extensions/memory_limiter/iotc_memory_limiter.c#L230-L232) removed?
